docs: boot script migration (#4362)
strongloop/loopback-next: docs/site/migration/boot-scripts.md

permalink: /doc/en/lb4/migration-boot-scripts.html
---

In LoopBack 3, predefined boot scripts are organized in the `/server/boot`
directory, and are executed right before the server starts to perform some
custom application initialization. The same functionality can also be achieved
in LoopBack 4 application by adding observers.

The LoopBack 3 boot script is mapped to LoopBack 4 observer's `start` function.
And moreover, besides participating in an application's start phase, an observer
also allows you to perform logic in the stop phase.

> ## Migrating Boot Scripts

Reviewer: Since this whole page is titled "Migrating boot scripts" (the level 1 heading), I feel it's redundant to introduce a level 2 heading with the same name.

I am proposing to remove this line and reduce level of the headings below, e.g. `## LoopBack 3 Boot Script`.

### LoopBack 3 Boot Script

> LoopBack 3 has different signatures for synchronized and asynchronized boot

Reviewer: AFAIK, we use the terms "synchronous" and "asynchronous", not synchronized and asynchronized. See e.g. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function.

```suggestion
LoopBack 3 has different signatures for synchronous and asynchronous boot
```

Please update all new content accordingly.

as `app`). A synchronized script only has one parameter, for example, the
following is a boot script that prints out a datasource's name before
application runs:

```js
// a typical synchronized boot script
// in file `server/boot/print-info.js`
'use strict';

module.exports = function printInfo(server) {
  // SUPPOSE your application has a datasource called `db`, which is usually the default
  // memory datasource in a sample LoopBack application created by CLI, regardless it's
  // an LB3 app or LB4 app
  const db = server.datasources.db;
  console.log('This is a synchronized script.');
  console.log('Your app has a datasource called: ', db.name);
};
```

While a typical asynchronized script has a callback function as the second
parameter:

```js
// a typical asynchronized boot script
// in file `server/boot/create-sample.js`
'use strict';

module.exports = function createSample(server, done) {
  // SUPPOSE your application has a Todo model
  const Todo = server.models.Todo;
  const sample = {title: 'a todo sample', desc: 'Something to do.'};
  Todo.create(sample)
    .then(result => {
      console.log('Sample created as: ', result);
    })
    .then(done);
};
```

> 

Reviewer: I believe that boot scripts can use `async` functions too (or in fact any function that returns a promise). Can you please mention this fact and perhaps show a code snippet too?

Also in the callback-based code snippet above, please don't mix promises with callbacks, the conversion from a promise to a callback is tricky. For example, your code above does not handle errors (promise rejections) and will trigger unhandled promise rejection warning.

A better example of a callback-based boot script:

```ts
module.exports = function createSample(server, done) {
  // SUPPOSE your application has a Todo model
  const Todo = server.models.Todo;
  const sample = {title: 'a todo sample', desc: 'Something to do.'};
  Todo.create(sample, (err, result) => {
    if (err) return done(err);
    console.log('Sample created as: ', result);
    done();
  });
 };
 ```

An example of a promise-based boot script:

```ts
module.exports = function createSample(server) {
  // SUPPOSE your application has a Todo model
  const Todo = server.models.Todo;
  const sample = {title: 'a todo sample', desc: 'Something to do.'};
  return Todo.create(sample)
    .then(result => {
      console.log('Sample created as: ', result);
    });
 };
 ```

Author: Good point, refactored this part to be:

- sync script
- async script with cb
- async script returning a promise

`server/boot/create-sample.js`) above as examples, let's first migrate
`server/boot/print-info.js` to your LoopBack 4 application.

#### Creating Observer

Run command `lb4 observer` to create an observer called `printInfo`:

```sh
lb4 observer
? Observer name: printInfo
? Observer group:
   create src/observers/print-info.observer.ts
   update src/observers/index.ts

Observer PrintInfo was created in src/observers/
```

You will find a file called `PrintInfo.observer.ts` created in folder
`src/observers/`. Open that file, there is a class called `PrintInfoObserver`,
which implements `LifeCycleObserver` and has `start` and `stop` functions. The
`start` function is where you should copy over the content of LB3's `printInfo`
function.

#### Injecting Application

The LoopBack 3 boot script takes in `server` as its first parameter, which is
essentially the `application` injected using key
`CoreBindings.APPLICATION_INSTANCE`:

```ts
// ... imports
export class PrintInfoObserver implements LifeCycleObserver {
  constructor(
    @inject(CoreBindings.APPLICATION_INSTANCE) private app: Application,
  ) {
    // ... other class members
  }

> ```

Reviewer: Please add `}` to close the class definition.

The markdown & typescript parser & highlighter used by GitHub is confused now and shows the text below in blue-ish color, while regular text should be rendered in black (see the text before this code snippet).

Author: oops

Now within the class `PrintInfoObserver` you have access to the application.

#### Migrating Artifacts

A boot script usually accesses or manipulates an application's artifacts like
models, datasources, etc... So when moving its content to the `start` function,
make sure you edit the code to retrieve those artifacts in LoopBack 4's way.

For the print info script, since it retrieves a datasource's info, you need to
**MIGRATE** the LoopBack datasource `db` to your LoopBack 4 application. The
steps are well documented on page [migration/datasources.md](./datasources.md).
And in the `start` function, **EDIT CODE** to retrieve the datasource using

> `this.app.getSync('datasources.db')`:

Reviewer: I have two comments:

1. In general, we should teach users to call `await app.get()`, because we cannot guarantee that all datasources will be always bound in a way that allows synchronous resolution.

2. Resolving dependencies manually via `get`/`getSync` is an anti-pattern. Users should use `@inject` instead. If the dependency is not bound at the time when constructor dependencies are resolved, then users should replace `@inject` with `@inject.getter`.

Author: Agreed changed to `@inject`
> because we cannot guarantee that all datasources will be always bound in a way that allows synchronous resolution.

Hmm...why it is not guaranteed? I think datasources are loaded first then the observers.

Reviewer: > > because we cannot guarantee that all datasources will be always bound in a way that allows synchronous resolution.
>
> Hmm...why it is not guaranteed? I think datasources are loaded first then the observers.

Consider the following code:

```ts
app
  .bind('datasources.db')
  .toDynamicValue(async function createDb() {
    const config = // do some async work
    return new juggler.DataSource(config);
  })
  .inScope(BindingScope.SINGLETON);
```

It binds the datasource in such way that the first call of `app.get` or `app.getSync` is required to perform async computation.

Author: I see thank you~

_The code snippet is the complete observer file, the LB3 boot script logic is
only in the `start` function_

```ts
import {
  /* inject, Application, CoreBindings, */
  lifeCycleObserver, // The decorator
  LifeCycleObserver,
  inject,
  CoreBindings,
  Application, // The interface
} from '@loopback/core';
import {juggler} from '@loopback/repository';

/**
 * This class will be bound to the application as a `LifeCycleObserver` during
 * `boot`
 */

> @lifeCycleObserver('')

Reviewer 2: Can't the `''` be omitted?

Author: yep, removed and added one more section for setting the scripts order.

`TodoRepository.create(sample)` instead to create the sample. To migrate the
`Todo` model, you can follow the steps in
[Migrating model definitions and built-in APIs](./models/core.md).

The migrated boot script as `CreateSampleObserver`'s start function:

_The code snippet is the complete observer file, the LB3 boot script logic is
only in the `start` function_

```ts
import {
  /* inject, Application, CoreBindings, */
  lifeCycleObserver, // The decorator
  LifeCycleObserver,
  inject,
  CoreBindings,
  Application, // The interface
} from '@loopback/core';
import {TodoRepository} from '../repositories';

/**
 * This class will be bound to the application as a `LifeCycleObserver` during
 * `boot`
 */
@lifeCycleObserver('')
export class CreateSampleObserver implements LifeCycleObserver {
  constructor(
    @inject(CoreBindings.APPLICATION_INSTANCE) private app: Application,
  ) {}

  /**
   * This method will be invoked when the application starts
   */
  async start(): Promise<void> {
    // Add your logic for start
    console.log('This is a migrated asynchronized boot script');

    // Retrieve the repository by key 'repositories.TodoRepository'

>     const TodoRepo: TodoRepository = this.app.getSync(

Reviewer: Please use `@inject` or `@inject.getter` to obtain the repository.

permalink: /doc/en/lb4/migration-boot-scripts.html
---

In LoopBack 3, predefined boot scripts are organized in the `/server/boot`
directory, and are executed right before the server starts to perform some
custom application initialization. The same functionality can also be achieved
in LoopBack 4 application by adding observers.

The LoopBack 3 boot script is mapped to LoopBack 4 observer's `start` function.
And moreover, besides participating in an application's start phase, an observer
also allows you to perform logic in the stop phase.

## LoopBack 3 Boot Script

LoopBack 3 supports synchronous and asynchronous boot scripts. And you can
choose either callback or promise based implementation for asynchronous scripts.

> They all have the first parameter as `server`(some people are used to name it as

Reviewer: Typo.

```suggestion
They all have the first parameter as `server` (some people are used to name it as
```

  lifeCycleObserver, // The decorator
  LifeCycleObserver,
  inject,
  CoreBindings,
  Application, // The interface
} from '@loopback/core';
import {TodoRepository} from '../repositories';

/**
 * This class will be bound to the application as a `LifeCycleObserver` during
 * `boot`
 */
@lifeCycleObserver()
export class CreateSampleObserver implements LifeCycleObserver {
  constructor(
    // inject `app` if you need access to other artifacts by `await this.app.get()`
    @inject(CoreBindings.APPLICATION_INSTANCE) private app: Application,
    // inject a repository with key `repositories.${repoName}`
    // or with the shortcut injector:
    // `@repository(TodoRepository) private todoRepo: TodoRepository`
    @inject('repositories.TodoRepository') private todoRepo: TodoRepository,
  ) {}

  /**
   * This method will be invoked when the application starts
   */
  async start(): Promise<void> {
    // Add your logic for start
    console.log('This is a migrated asynchronous boot script');

    const sample = {title: 'a todo sample', desc: 'Something to do.'};
    // Create the sample by calling TodoRepo.create()
    const result = await TodoRepo.create(sample);

>     console.log(`Sample created as: ${result}`);

Reviewer: AFAICT from the CI logs, this line prints the following output:

```
Sample created as: [object Object]
```

`[object Object]` is not very useful. I am proposing to rewrite this line as follows:

```suggestion
    console.log('Sample created as', result);
```

  }

  /**
   * This method will be invoked when the application stops
   */
  async stop(): Promise<void> {
    // Add your logic for stop
    console.log('create sample observer has stopped.');
  }
}
```

### Setting Order for Migrated Boot Scripts

In LoopBack 3, boot scripts are loaded according to the script file'

> name(alphabetically). LoopBack 4 observer provides an advanced way to control

Reviewer: ```suggestion
name (alphabetically). LoopBack 4 observer provides an advanced way to control
```

the order of scripts by specifying and setting group names. The tutorial is well
documented in the section
[Notify life cycle observers of start/stop related events by order](../Life-cycle.md#notify-life-cycle-observers-of-startstop-related-events-by-order).

## Predefined LoopBack 3 Boot Scripts

LoopBack 3 application has two predefined boot scripts: `/server/boot/root.js`
and `/server/boot/authentication.js`. Please do not create corresponding

> observers for them because:

Reviewer: I feel the word "because" is redundant here.

```suggestion
observers for them.
```

permalink: /doc/en/lb4/migration-boot-scripts.html
---

In LoopBack 3, predefined boot scripts are organized in the `/server/boot`
directory, and are executed right before the server starts to perform some
custom application initialization. The same functionality can also be achieved
in LoopBack 4 application by adding observers.

The LoopBack 3 boot script is mapped to LoopBack 4 observer's `start` function.
And moreover, besides participating in an application's start phase, an observer
also allows you to perform logic in the stop phase.

## LoopBack 3 Boot Script

LoopBack 3 supports synchronous and asynchronous boot scripts. And you can
choose either callback or promise based implementation for asynchronous scripts.
They all have the first parameter as `server` (some people are used to name it
as `app`). The callback based asynchronous script has a 2nd parameter `done` as
the callback function.

> A typical synchronous boot script that prints out a datasource's name before

Reviewer 3: Maybe say `LoopBack 3 supports three styles of boot scripts` and make each of the examples below a bullet item.

 */
@lifeCycleObserver()
export class CreateSampleObserver implements LifeCycleObserver {
  constructor(
    // inject `app` if you need access to other artifacts by `await this.app.get()`
    @inject(CoreBindings.APPLICATION_INSTANCE) private app: Application,
    // inject a repository with key `repositories.${repoName}`
    // or with the shortcut injector:
    // `@repository(TodoRepository) private todoRepo: TodoRepository`
    @inject('repositories.TodoRepository') private todoRepo: TodoRepository,
  ) {}

  /**
   * This method will be invoked when the application starts
   */
  async start(): Promise<void> {
    // Add your logic for start
    console.log('This is a migrated asynchronous boot script');

    const sample = {title: 'a todo sample', desc: 'Something to do.'};
    // Create the sample by calling TodoRepo.create()
    const result = await TodoRepo.create(sample);
    console.log('Sample created as ', result);
  }

  /**
   * This method will be invoked when the application stops
   */
  async stop(): Promise<void> {
    // Add your logic for stop
    console.log('create sample observer has stopped.');
  }
}
```

### Setting Order for Migrated Boot Scripts

> In LoopBack 3, boot scripts are loaded according to the script file' name

Reviewer 4: `file'` -> `file's`?

Author: good catch
